Performance issue on large project

I am trying to integrate Java language server to my vscode auto-completion extension. I found that there exists performance issue in eclipse.jdt on large project. I tried to start the server on a Java project contains more than 3000 files. In the first few minutes, the server did not response the `textDocument/completion` request, and then crashed because OOM:
```
[Info  - 1:12:53 PM] Jun 4, 2020 1:12:53 PM >> document/completion
[Info  - 1:12:53 PM] Jun 4, 2020 1:12:53 PM Reconciled 0. Took 111 ms
[Error - 1:18:35 PM] Request textDocument/completion failed.
  Message: Internal error.
  Code: -32603 
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$0(ParentProcessWatcher.java:123)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 12 more
Caused by: java.util.concurrent.CompletionException: java.lang.OutOfMemoryError: Java heap space
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:443)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOfRange(Arrays.java:3664)
	at java.lang.String.<init>(String.java:207)
	at java.lang.StringBuilder.toString(StringBuilder.java:407)
	at org.eclipse.jface.text.GapTextStore.get(GapTextStore.java:165)
	at org.eclipse.jface.text.CopyOnWriteTextStore.get(CopyOnWriteTextStore.java:128)
	at org.eclipse.jface.text.AbstractDocument.get(AbstractDocument.java:961)
	at org.eclipse.core.internal.filebuffers.SynchronizableDocument.get(SynchronizableDocument.java:119)
	at org.eclipse.jdt.ls.core.internal.DocumentAdapter.getContents(DocumentAdapter.java:190)
	at org.eclipse.jdt.internal.core.Openable.findRecommendedLineSeparator(Openable.java:219)
	at org.eclipse.jdt.internal.core.dom.rewrite.imports.ImportRewriteAnalyzer.<init>(ImportRewriteAnalyzer.java:487)
	at org.eclipse.jdt.core.dom.rewrite.ImportRewrite.rewriteImports(ImportRewrite.java:1328)
	at org.eclipse.jdt.ls.core.internal.contentassist.CompletionProposalReplacementProvider.addImports(CompletionProposalReplacementProvider.java:354)
	at org.eclipse.jdt.ls.core.internal.contentassist.CompletionProposalReplacementProvider.updateReplacement(CompletionProposalReplacementProvider.java:177)
	at org.eclipse.jdt.ls.core.internal.contentassist.CompletionProposalRequestor.toCompletionItem(CompletionProposalRequestor.java:268)
	at org.eclipse.jdt.ls.core.internal.contentassist.CompletionProposalRequestor.getCompletionItems(CompletionProposalRequestor.java:206)
	at org.eclipse.jdt.ls.core.internal.handlers.CompletionHandler.computeContentAssist(CompletionHandler.java:147)
	at org.eclipse.jdt.ls.core.internal.handlers.CompletionHandler.completion(CompletionHandler.java:76)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$4(JDTLanguageServer.java:494)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer$$Lambda$155/780388676.apply(Unknown Source)
	at org.eclipse.jdt.ls.core.internal.BaseJDTLanguageServer.lambda$0(BaseJDTLanguageServer.java:74)
	at org.eclipse.jdt.ls.core.internal.BaseJDTLanguageServer$$Lambda$151/788893917.apply(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
	... 6 more

```

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/HoverInfoProvider.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/HoverInfoProvider.java
@@ -91,10 +91,12 @@ public class HoverInfoProvider {
 	public List<Either<String, MarkedString>> computeHover(int line, int column, IProgressMonitor monitor) {
 		List<Either<String, MarkedString>> res = new LinkedList<>();
 		try {
+			if (monitor.isCanceled()) {
+				return cancelled(res);
+			}
 			IJavaElement[] elements = JDTUtils.findElementsAtSelection(unit, line, column, this.preferenceManager, monitor);
-			if(elements == null || elements.length == 0) {
-				res.add(Either.forLeft(""));
-				return res;
+			if (elements == null || elements.length == 0 || monitor.isCanceled()) {
+				return cancelled(res);
 			}
 			IJavaElement curr = null;
 			if (elements.length != 1) {
@@ -116,7 +118,9 @@ public class HoverInfoProvider {
 			} else {
 				curr = elements[0];
 			}
-
+			if (monitor.isCanceled()) {
+				return cancelled(res);
+			}
 			if (JDTEnvironmentUtils.isSyntaxServer() || isResolved(curr, monitor)) {
 				IBuffer buffer = curr.getOpenable().getBuffer();
 				if (buffer == null && curr instanceof BinaryMember) {
@@ -128,9 +132,15 @@ public class HoverInfoProvider {
 						}
 					}
 				}
+				if (monitor.isCanceled()) {
+					return cancelled(res);
+				}
 				MarkedString signature = computeSignature(curr);
 				if (signature != null) {
 					res.add(Either.forRight(signature));
+				}
+				if (monitor.isCanceled()) {
+					return cancelled(res);
 				}
 				MarkedString javadoc = computeJavadoc(curr);
 				if (javadoc != null && javadoc.getValue() != null) {
@@ -140,6 +150,15 @@ public class HoverInfoProvider {
 		} catch (Exception e) {
 			JavaLanguageServerPlugin.logException("Error computing hover", e);
 		}
+		if (monitor.isCanceled()) {
+			return cancelled(res);
+		}
+		return res;
+	}
+
+	private List<Either<String, MarkedString>> cancelled(List<Either<String, MarkedString>> res) {
+		res.clear();
+		res.add(Either.forLeft(""));
 		return res;
 	}
 

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java
@@ -45,6 +45,7 @@ import org.eclipse.core.runtime.OperationCanceledException;
 import org.eclipse.core.runtime.Path;
 import org.eclipse.core.runtime.Platform;
 import org.eclipse.core.runtime.URIUtil;
+import org.eclipse.core.runtime.jobs.ISchedulingRule;
 import org.eclipse.jdt.core.CompletionProposal;
 import org.eclipse.jdt.core.Flags;
 import org.eclipse.jdt.core.IAnnotatable;
@@ -703,14 +704,14 @@ public final class JDTUtils {
 
 	public static IJavaElement findElementAtSelection(ITypeRoot unit, int line, int column, PreferenceManager preferenceManager, IProgressMonitor monitor) throws JavaModelException {
 		IJavaElement[] elements = findElementsAtSelection(unit, line, column, preferenceManager, monitor);
-		if (elements != null && elements.length == 1) {
+		if ((elements != null && elements.length == 1) || monitor.isCanceled()) {
 			return elements[0];
 		}
 		return null;
 	}
 
 	public static IJavaElement[] findElementsAtSelection(ITypeRoot unit, int line, int column, PreferenceManager preferenceManager, IProgressMonitor monitor) throws JavaModelException {
-		if (unit == null) {
+		if (unit == null || monitor.isCanceled()) {
 			return null;
 		}
 		int offset = JsonRpcHelpers.toOffset(unit.getBuffer(), line, column);
@@ -799,6 +800,14 @@ public final class JDTUtils {
 
 	public static IFile findFile(String uriString) {
 		return (IFile) findResource(toURI(uriString), ResourcesPlugin.getWorkspace().getRoot()::findFilesForLocationURI);
+	}
+
+	public static ISchedulingRule getRule(String uri) {
+		IResource resource = JDTUtils.findFile(uri);
+		if (resource != null) {
+			return ResourcesPlugin.getWorkspace().getRuleFactory().createRule(resource);
+		}
+		return null;
 	}
 
 	public static IContainer findFolder(String uriString) {

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDocumentLifeCycleHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDocumentLifeCycleHandler.java
@@ -22,6 +22,8 @@ import java.util.Set;
 
 import org.eclipse.core.resources.IFile;
 import org.eclipse.core.resources.IResource;
+import org.eclipse.core.resources.IResourceRuleFactory;
+import org.eclipse.core.resources.IWorkspace;
 import org.eclipse.core.resources.IWorkspaceRunnable;
 import org.eclipse.core.resources.ResourcesPlugin;
 import org.eclipse.core.resources.WorkspaceJob;
@@ -32,6 +34,8 @@ import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.NullProgressMonitor;
 import org.eclipse.core.runtime.Status;
 import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.core.runtime.jobs.ISchedulingRule;
+import org.eclipse.core.runtime.jobs.MultiRule;
 import org.eclipse.jdt.core.IBuffer;
 import org.eclipse.jdt.core.ICompilationUnit;
 import org.eclipse.jdt.core.IJavaElement;
@@ -90,7 +94,6 @@ public abstract class BaseDocumentLifeCycleHandler {
 					return DOCUMENT_LIFE_CYCLE_JOBS.equals(family);
 				}
 			};
-			this.validationTimer.setRule(ResourcesPlugin.getWorkspace().getRoot());
 			this.publishDiagnosticsJob = new WorkspaceJob("Publish Diagnostics") {
 				@Override
 				public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
@@ -105,7 +108,6 @@ public abstract class BaseDocumentLifeCycleHandler {
 					return PUBLISH_DIAGNOSTICS_JOBS.equals(family);
 				}
 			};
-			this.validationTimer.setRule(ResourcesPlugin.getWorkspace().getRoot());
 		}
 	}
 
@@ -129,30 +131,53 @@ public abstract class BaseDocumentLifeCycleHandler {
 		}
 		if (validationTimer != null) {
 			validationTimer.cancel();
+			ISchedulingRule rule = getRule(toReconcile);
 			if (publishDiagnosticsJob != null) {
 				publishDiagnosticsJob.cancel();
+				publishDiagnosticsJob.setRule(rule);
 			}
+			validationTimer.setRule(rule);
 			validationTimer.schedule(delay);
 		} else {
 			performValidation(new NullProgressMonitor());
 		}
 	}
 
+	private ISchedulingRule getRule(Set<ICompilationUnit> units) {
+		ISchedulingRule result = null;
+		IResourceRuleFactory ruleFactory = ResourcesPlugin.getWorkspace().getRuleFactory();
+		for (ICompilationUnit unit : units) {
+			if (unit.getResource() != null) {
+				ISchedulingRule rule = ruleFactory.createRule(unit.getResource());
+				result = MultiRule.combine(rule, result);
+			}
+		}
+		return result;
+	}
+
 	private IStatus performValidation(IProgressMonitor monitor) throws JavaModelException {
 		long start = System.currentTimeMillis();
 
-		List<ICompilationUnit> cusToReconcile = new ArrayList<>();
+		List<ICompilationUnit> cusToReconcile;
 		synchronized (toReconcile) {
+			if (toReconcile.isEmpty()) {
+				return Status.OK_STATUS;
+			}
+			cusToReconcile = new ArrayList<>(toReconcile.size());
 			cusToReconcile.addAll(toReconcile);
 			toReconcile.clear();
 		}
-		if (cusToReconcile.isEmpty()) {
-			return Status.OK_STATUS;
+		if (monitor.isCanceled()) {
+			return Status.CANCEL_STATUS;
 		}
 		// first reconcile all units with content changes
 		SubMonitor progress = SubMonitor.convert(monitor, cusToReconcile.size() + 1);
 		for (ICompilationUnit cu : cusToReconcile) {
-			cu.reconcile(ICompilationUnit.NO_AST, true, null, progress.newChild(1));
+			if (monitor.isCanceled()) {
+				return Status.CANCEL_STATUS;
+			}
+			cu.makeConsistent(progress);
+			//cu.reconcile(ICompilationUnit.NO_AST, false, null, progress.newChild(1));
 		}
 		JavaLanguageServerPlugin.logInfo("Reconciled " + toReconcile.size() + ". Took " + (System.currentTimeMillis() - start) + " ms");
 		if (monitor.isCanceled()) {
@@ -164,6 +189,9 @@ public abstract class BaseDocumentLifeCycleHandler {
 				publishDiagnosticsJob.join();
 			} catch (InterruptedException e) {
 				// ignore
+			}
+			if (monitor.isCanceled()) {
+				return Status.CANCEL_STATUS;
 			}
 			publishDiagnosticsJob.schedule(400);
 		} else {
@@ -179,6 +207,9 @@ public abstract class BaseDocumentLifeCycleHandler {
 		}
 		this.sharedASTProvider.disposeAST();
 		List<ICompilationUnit> toValidate = Arrays.asList(JavaCore.getWorkingCopies(null));
+		if (toValidate.isEmpty()) {
+			return Status.OK_STATUS;
+		}
 		SubMonitor progress = SubMonitor.convert(monitor, toValidate.size() + 1);
 		if (monitor.isCanceled()) {
 			return Status.CANCEL_STATUS;
@@ -188,6 +219,9 @@ public abstract class BaseDocumentLifeCycleHandler {
 				return Status.CANCEL_STATUS;
 			}
 			CompilationUnit astRoot = this.sharedASTProvider.getAST(rootToValidate, CoreASTProvider.WAIT_YES, monitor);
+			if (monitor.isCanceled()) {
+				return Status.CANCEL_STATUS;
+			}
 			if (astRoot != null) {
 				// report errors, even if there are no problems in the file: The client need to know that they got fixed.
 				ICompilationUnit unit = (ICompilationUnit) astRoot.getTypeRoot();
@@ -207,10 +241,12 @@ public abstract class BaseDocumentLifeCycleHandler {
 			 */
 			@Override
 			public IBuffer createBuffer(ICompilationUnit workingCopy) {
-				ICompilationUnit original = workingCopy.getPrimary();
-				IResource resource = original.getResource();
-				if (resource instanceof IFile) {
-					return new DocumentAdapter(workingCopy, (IFile) resource);
+				if (!monitor.isCanceled()) {
+					ICompilationUnit original = workingCopy.getPrimary();
+					IResource resource = original.getResource();
+					if (resource instanceof IFile) {
+						return new DocumentAdapter(workingCopy, (IFile) resource);
+					}
 				}
 				return DocumentAdapter.Null;
 			}
@@ -229,45 +265,49 @@ public abstract class BaseDocumentLifeCycleHandler {
 	}
 
 	public void didClose(DidCloseTextDocumentParams params) {
+		ISchedulingRule rule = JDTUtils.getRule(params.getTextDocument().getUri());
 		try {
 			ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
 				@Override
 				public void run(IProgressMonitor monitor) throws CoreException {
 					handleClosed(params);
 				}
-			}, new NullProgressMonitor());
+			}, rule, IWorkspace.AVOID_UPDATE, new NullProgressMonitor());
 		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException("Handle document close ", e);
 		}
 	}
 
 	public void didOpen(DidOpenTextDocumentParams params) {
+		ISchedulingRule rule = JDTUtils.getRule(params.getTextDocument().getUri());
 		try {
 			ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
 				@Override
 				public void run(IProgressMonitor monitor) throws CoreException {
 					handleOpen(params);
 				}
-			}, new NullProgressMonitor());
+			}, rule, IWorkspace.AVOID_UPDATE, new NullProgressMonitor());
 		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException("Handle document open ", e);
 		}
 	}
 
 	public void didChange(DidChangeTextDocumentParams params) {
+		ISchedulingRule rule = JDTUtils.getRule(params.getTextDocument().getUri());
 		try {
 			ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
 				@Override
 				public void run(IProgressMonitor monitor) throws CoreException {
 					handleChanged(params);
 				}
-			}, new NullProgressMonitor());
+			}, rule, IWorkspace.AVOID_UPDATE, new NullProgressMonitor());
 		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException("Handle document change ", e);
 		}
 	}
 
 	public void didSave(DidSaveTextDocumentParams params) {
+		ISchedulingRule rule = JDTUtils.getRule(params.getTextDocument().getUri());
 		try {
 			JobHelpers.waitForJobs(DocumentLifeCycleHandler.DOCUMENT_LIFE_CYCLE_JOBS, new NullProgressMonitor());
 			ResourcesPlugin.getWorkspace().run(new IWorkspaceRunnable() {
@@ -275,7 +315,7 @@ public abstract class BaseDocumentLifeCycleHandler {
 				public void run(IProgressMonitor monitor) throws CoreException {
 					handleSaved(params);
 				}
-			}, new NullProgressMonitor());
+			}, rule, IWorkspace.AVOID_UPDATE, new NullProgressMonitor());
 		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException("Handle document save ", e);
 		}

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/CodeActionHandler.java
@@ -77,16 +77,20 @@ public class CodeActionHandler {
 	}
 
 	public List<Either<Command, CodeAction>> getCodeActionCommands(CodeActionParams params, IProgressMonitor monitor) {
-		final ICompilationUnit unit = JDTUtils.resolveCompilationUnit(params.getTextDocument().getUri());
-		if (unit == null) {
+		if (monitor.isCanceled()) {
 			return Collections.emptyList();
 		}
-
+		final ICompilationUnit unit = JDTUtils.resolveCompilationUnit(params.getTextDocument().getUri());
+		if (unit == null || monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 		int start = DiagnosticsHelper.getStartOffset(unit, params.getRange());
 		int end = DiagnosticsHelper.getEndOffset(unit, params.getRange());
 		InnovationContext context = new InnovationContext(unit, start, end - start);
 		context.setASTRoot(getASTRoot(unit));
-
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 		IProblemLocationCore[] locations = this.getProblemLocationCores(unit, params.getContext().getDiagnostics());
 
 		List<String> codeActionKinds = new ArrayList<>();
@@ -115,6 +119,9 @@ public class CodeActionHandler {
 				JavaLanguageServerPlugin.logException("Problem resolving quick fix code actions", e);
 			}
 		}
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 
 		if (containsKind(codeActionKinds, CodeActionKind.Refactor)) {
 			try {
@@ -125,7 +132,9 @@ public class CodeActionHandler {
 				JavaLanguageServerPlugin.logException("Problem resolving refactor code actions", e);
 			}
 		}
-
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 		if (containsKind(codeActionKinds, JavaCodeActionKind.QUICK_ASSIST)) {
 			try {
 				List<ChangeCorrectionProposal> quickassistProposals = this.quickAssistProcessor.getAssists(params, context, locations);
@@ -135,7 +144,9 @@ public class CodeActionHandler {
 				JavaLanguageServerPlugin.logException("Problem resolving quick assist code actions", e);
 			}
 		}
-
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 		try {
 			for (ChangeCorrectionProposal proposal : proposals) {
 				Optional<Either<Command, CodeAction>> codeActionFromProposal = getCodeActionFromProposal(proposal, params.getContext());
@@ -146,9 +157,14 @@ public class CodeActionHandler {
 		} catch (CoreException e) {
 			JavaLanguageServerPlugin.logException("Problem converting proposal to code actions", e);
 		}
-
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 		if (containsKind(codeActionKinds, CodeActionKind.Source)) {
 			codeActions.addAll(sourceAssistProcessor.getSourceActionCommands(params, context, locations));
+		}
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
 		}
 		return codeActions;
 	}

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentLifeCycleHandler.java
@@ -18,8 +18,6 @@ import java.util.Collection;
 import java.util.List;
 import java.util.Optional;
 
-import com.google.common.collect.Iterables;
-
 import org.eclipse.core.resources.IFile;
 import org.eclipse.core.runtime.IPath;
 import org.eclipse.jdt.core.ICompilationUnit;
@@ -51,6 +49,8 @@ import org.eclipse.text.edits.InsertEdit;
 import org.eclipse.text.edits.MalformedTreeException;
 import org.eclipse.text.edits.ReplaceEdit;
 import org.eclipse.text.edits.TextEdit;
+
+import com.google.common.collect.Iterables;
 
 public class DocumentLifeCycleHandler extends BaseDocumentLifeCycleHandler {
 
@@ -86,19 +86,20 @@ public class DocumentLifeCycleHandler extends BaseDocumentLifeCycleHandler {
 			unit = JDTUtils.resolveCompilationUnit(resource);
 		} else { // Open the standalone files.
 			IPath filePath = ResourceUtils.canonicalFilePathFromURI(uri);
-			Collection<IPath> rootPaths = preferenceManager.getPreferences().getRootPaths();
-			Optional<IPath> belongedRootPath = rootPaths.stream().filter(rootPath -> rootPath.isPrefixOf(filePath)).findFirst();
-			boolean invisibleProjectEnabled = false;
-			if (belongedRootPath.isPresent()) {
-				IPath rootPath = belongedRootPath.get();
-				invisibleProjectEnabled = InvisibleProjectImporter.loadInvisibleProject(filePath, rootPath, false);
-				if (invisibleProjectEnabled) {
-					unit = JDTUtils.resolveCompilationUnit(uri);
+			if (filePath != null) {
+				Collection<IPath> rootPaths = preferenceManager.getPreferences().getRootPaths();
+				Optional<IPath> belongedRootPath = rootPaths.stream().filter(rootPath -> rootPath.isPrefixOf(filePath)).findFirst();
+				boolean invisibleProjectEnabled = false;
+				if (belongedRootPath.isPresent()) {
+					IPath rootPath = belongedRootPath.get();
+					invisibleProjectEnabled = InvisibleProjectImporter.loadInvisibleProject(filePath, rootPath, false);
+					if (invisibleProjectEnabled) {
+						unit = JDTUtils.resolveCompilationUnit(uri);
+					}
 				}
-			}
-
-			if (!invisibleProjectEnabled) {
-				unit = JDTUtils.getFakeCompilationUnit(uri);
+				if (!invisibleProjectEnabled) {
+					unit = JDTUtils.getFakeCompilationUnit(uri);
+				}
 			}
 		}
 
@@ -193,6 +194,7 @@ public class DocumentLifeCycleHandler extends BaseDocumentLifeCycleHandler {
 		return unit;
 	}
 
+	@Override
 	public ICompilationUnit handleClosed(DidCloseTextDocumentParams params) {
 		ICompilationUnit unit = super.handleClosed(params);
 		uninstallSemanticHighlightings(params.getTextDocument().getUri());

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FindLinksHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/FindLinksHandler.java
@@ -34,16 +34,19 @@ import org.eclipse.lsp4j.TextDocumentPositionParams;
 public class FindLinksHandler {
 
 	public static List<? extends Location> findLinks(String linkType, TextDocumentPositionParams position, IProgressMonitor monitor) {
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 		ITypeRoot unit = JDTUtils.resolveTypeRoot(position.getTextDocument().getUri());
 		if (unit != null && !monitor.isCanceled()) {
 			PreferenceManager preferenceManager = JavaLanguageServerPlugin.getInstance().getPreferencesManager();
 			try {
 				IJavaElement element = JDTUtils.findElementAtSelection(unit, position.getPosition().getLine(), position.getPosition().getCharacter(), preferenceManager, monitor);
-				if (Objects.equals(linkType, "superImplementation")) {
+				if (!monitor.isCanceled() && Objects.equals(linkType, "superImplementation")) {
 					IMethod overriddenMethod = findOverriddenMethod(element, monitor);
-					if (overriddenMethod != null) {
+					if (!monitor.isCanceled() && overriddenMethod != null) {
 						Location location = NavigateToDefinitionHandler.computeDefinitionNavigation(overriddenMethod, element.getJavaProject());
-						if (location != null) {
+						if (!monitor.isCanceled() && location != null) {
 							String declaringTypeName = overriddenMethod.getDeclaringType().getFullyQualifiedName();
 							String methodName = overriddenMethod.getElementName();
 							String displayName = declaringTypeName + "." + methodName;

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/NavigateToDefinitionHandler.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/NavigateToDefinitionHandler.java
@@ -60,18 +60,24 @@ public class NavigateToDefinitionHandler {
 	}
 
 	public List<? extends Location> definition(TextDocumentPositionParams position, IProgressMonitor monitor) {
+		if (monitor.isCanceled()) {
+			return Collections.emptyList();
+		}
 		ITypeRoot unit = JDTUtils.resolveTypeRoot(position.getTextDocument().getUri());
 		Location location = null;
 		if (unit != null && !monitor.isCanceled()) {
 			location = computeDefinitionNavigation(unit, position.getPosition().getLine(),
 					position.getPosition().getCharacter(), monitor);
 		}
-		return location == null ? Collections.emptyList() : Arrays.asList(location);
+		return location == null || monitor.isCanceled() ? Collections.emptyList() : Arrays.asList(location);
 	}
 
 	private Location computeDefinitionNavigation(ITypeRoot unit, int line, int column, IProgressMonitor monitor) {
 		try {
 			IJavaElement element = JDTUtils.findElementAtSelection(unit, line, column, this.preferenceManager, monitor);
+			if (monitor.isCanceled()) {
+				return null;
+			}
 			if (element == null) {
 				return computeBreakContinue(unit, line, column);
 			}

